File: TypeScript/Linked List/DoublyLinkedList.ts
class DoublyNode {
  data: number;
  next: DoublyNode | null;
  prev: DoublyNode | null;

  constructor(data: number) {
    this.data = data;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  head: DoublyNode | null;
  tail: DoublyNode | null;

  constructor() {
    this.head = null;
    this.tail = null;
  }

  add(data: number): void {
    const newNode = new DoublyNode(data);
    if (this.head === null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail!.next = newNode;
      newNode.prev = this.tail;
      this.tail = newNode;
    }
  }

  search(data: number): boolean {
    let current = this.head;
    while (current !== null) {
      if (current.data === data) {
        return true;
      }
      current = current.next;
    }
    return false;
  }

  delete(data: number): void {
    if (this.head === null) {
      return;
    }

    if (this.head.data === data) {
      this.head = this.head.next;
      if (this.head !== null) {
        this.head.prev = null;
      } else {
        this.tail = null;
      }
      return;
    }

    let current: DoublyNode | null = this.head;
    while (current !== null && current.data !== data) {
      current = current.next;
    }

    if (current === null) {
      return;
    }

    if (current.next !== null) {
      current.next.prev = current.prev;
    } else {
      this.tail = current.prev;
    }

    if (current.prev !== null) {
      current.prev.next = current.next;
    }
  }

  show(): void {
    let current = this.head;
    while (current !== null) {
      console.log(`The address of node is ${current}.`);
      console.log(`The data is ${current.data}.`);
      console.log(`The address of next node is ${current.next}.`);
      console.log(`The address of previous node is ${current.prev}.`);
      console.log(" ");
      current = current.next;
    }
  }
}

let doublyLinkedList = new DoublyLinkedList();
doublyLinkedList.add(1);
doublyLinkedList.add(3);
doublyLinkedList.add(2);
doublyLinkedList.add(4);
doublyLinkedList.delete(2);
doublyLinkedList.show();

File: Java/Final Term/Linked List/DoublyLinkedList.java
class Node {
    int data;
    Node next;
    Node prev;

    Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

public class DoublyLinkedList {
    public Node head;
    public Node tail;

    DoublyLinkedList() {
        this.head = null;
        this.tail = null;
    }

    void add(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            newNode.prev = tail;
            tail = newNode;
        }
    }

    boolean search(int data) {
        if (head == null) {
            return false;
        }
        Node current = head;
        while (current != null) {
            if (current.data == data) {
                return true;
            }
            current = current.next;
        }
        return false;
    }

    void showDataFromHead() {
        Node current = head;
        while (current != null) {
            System.out.println(current.data);
            current = current.next;
        }
    }

    void showDataFromTail() {
        Node current = tail;
        while (current != null) {
            System.out.println(current.data);
            current = current.prev;
        }
    }

    void delete(int data) {
        if (head == null) {
            return;
        }
        if (head.data == data) {
            head = head.next;
            if (head != null) {
                head.prev = null;
            } else {
                tail = null;
            }
            return;
        }
        Node current = head;
        while (current != null && current.data != data) {
            current = current.next;
        }
        if (current == null) {
            return;
        }
        if (current.next != null) {
            current.next.prev = current.prev;
        } else {
            tail = current.prev;
        }
        if (current.prev != null) {
            current.prev.next = current.next;
        }
    }

    public static void main(String[] args) {
        DoublyLinkedList list = new DoublyLinkedList();
        list.add(10);
        list.add(20);
        list.add(30);
        list.add(40);
        list.delete(30);
        list.showDataFromHead();
    }
}
